### ISSUE TITLE
Improved UX (error message) for generator users missing partial modifiers

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/roslyn#61597]
So, this is an issue that keeps happening, and we're getting lots of issues in [our own repo](https://github.com/CommunityToolkit/dotnet) (for the MVVM Toolkit source generator) from people trying out our generators, forgetting the `partial` modifier (or simply not knowing it's needed or that it even exists at all), and then opening issues thinking our generators are broken. For instance, see: https://github.com/CommunityToolkit/dotnet/issues/286.

This to me seems like a very, very common problem that pretty much all authors of source generators creating members in existing types will have to face, so I'm wondering what the correct solution in the long run might be. Wanted to get some thoughts from folks here before going ahead and doing extra work in the MVVM Toolkit to eg. emit additional diagnostics there.

There's multiple ways we could go about this, that I can imagine:

- Each source generator author should implement additional logic to detect whether a target type (including any types in the tree if the type is nested) are missing the `partial` modifier, and (1) emit a specific diagnostic and (2) maybe stop generation there.
- Treat all partial types in source generators as not needing partial in their corresponding type in the project. This has been suggested before but has some drawbacks, namely that it would hide from authors which types are receiving additional generated members.
- Have Roslyn detect cases like this (ie. a non-partial user type that has a partial version generated by a referenced generator), and emit a better error message, informing the user that they're missing a partial modifier, that the colliding file in the source generator needs that, etc.

To me, option 3 seems like the one that might require the least work and that will benefit everyone.

Note, we already do get a similar diagnostic, specifically this:

![](https://user-images.githubusercontent.com/10199417/171190888-96cc8017-61bd-41d7-b487-e57dcb2a0497.png)

But apparently this is not really enough, since people keep stumbling over this and reporting issues.

Would it maybe be possible to customize this error message when Roslyn can see that the partial declaration is from a source generator, to try to give clearer and more helpful instructions to the user (especially if they're beginners)?

**Comment by CyrusNajmabadi** [comment]
@Sergio0694 yes. The error message could likely be made better. Could you potentially submit a PR on this?

**Comment by mrlacey** [comment]
While I don't want to try and argue for more work than is needed, I worry that the underlying issue here is based on people not reading the existing error messages. Will altering/improving the messages make a sufficient difference?
I guess it's going to be necessary to try this and see if it helps. More speculation isn't going to help without actual data.

**Comment by Neme12** [comment]
> While I don't want to try and argue for more work than is needed, I worry that the underlying issue here is based on people not reading the existing error messages. 

The issue could also be that people do read the error messages, but they don't know why they should add partial, because they didn't make the class partial in any of their code and not know that the source generator needs that or that it's doing that for this specific class.

**Comment by CyrusNajmabadi** [comment]
I'd argue the generator should report a custom error here and say that the type they're augmenting needs to have the partial modifier on it

**Comment by Sergio0694** [comment]
Doesn't it seem like a lot of churn and also inevitably inconsistent behavior across all existing generators if we just decide the right approach here is to ask every single source generator author to do this check and emit a custom diagnostic themselves? 😟

**Comment by mrlacey** [comment]
> Doesn't it seem like a lot of churn and also inevitably inconsistent behavior across all existing generators if we just decide the right approach here is to ask every single source generator author to do this check and emit a custom diagnostic themselves? 😟

There's a counterargument that it's potentially better to encourage source generator authors to think about all their assumptions and ensure that they provide specific, appropriate messaging when they're not being met.
But if it's simple to also handle the case of a source generator outputting a partial class when the original wasn't partial and the compiler can detect this and add a suitable message that points to a solution in a way that is better than the current behavior then doing that as well is also good.

**Comment by Neme12** [comment]
@CyrusNajmabadi 
On Discord, @333fred suggested that this should work for any files that are considered generated, ie. not just from source generators, but also files that end with .g.cs or have the \<auto-generated\> comment, so that this would provide a helpful error message even in eg. WinForms, where they can't provide any custom diagnostics, but it could still be useful for the compiler to tell the user that there's another partial declaration in a generated file.

**Comment by mrlacey** [comment]
> @CyrusNajmabadi On Discord, @333fred suggested that this should work for any files that are considered generated, ie. not just from source generators, but also files that end with .g.cs or have the <auto-generated> comment, so that this would provide a helpful error message even in eg. WinForms, where they can't provide any custom diagnostics, but it could still be useful for the compiler to tell the user that there's another partial declaration in a generated file that the user might not know about.

What should be reported in that case?
The compiler already warns that some definitions are marked partial but others aren't. The OP here was to add a different (better) message when the ones marked as partial can be identified as coming from a source generator. What are the other cases where a specific (different) error message would be helpful? and what should the new message be?

**Comment by Neme12** [comment]
> > @CyrusNajmabadi On Discord, @333fred suggested that this should work for any files that are considered generated, ie. not just from source generators, but also files that end with .g.cs or have the <auto-generated> comment, so that this would provide a helpful error message even in eg. WinForms, where they can't provide any custom diagnostics, but it could still be useful for the compiler to tell the user that there's another partial declaration in a generated file that the user might not know about.
> 
> What should be reported in that case?
> The compiler already warns that some definitions are marked partial but others aren't. The OP here was to add a different (better) message when the ones marked as partial can be identified as coming from a source generator. What are the other cases where a specific (different) error message would be helpful? and what should the new message be?

It would be the same error message as when the partial declaration is coming from a source generator, but the error message wouldn't mention roslyn source generators specifically but rather that it's from a generated file I guess.

**Comment by Neme12** [comment]
Although yeah I agree that if it the message could mention source generators specifically, it could be more helpful to the user.

**Comment by sharwell** [comment]
If we just provide a code fix for this case that adds the `partial` modifier, it would be easy for the user to correct it. Also, we should make sure that if the user hits <kbd>F12</kbd> at the location of the error, they see all files defining the type.

> I'd argue the generator should report a custom error here and say that the type they're augmenting needs to have the partial modifier on it

The guidance I've been giving is source generators generally do not need to provide a diagnostic for any case where the compiler is already going to report a compilation error.

**Comment by Sergio0694** [comment]
Yeah I remember us discussing this a while back and that was the guidance I received, and why I never implemented such checks in none of my source generators (eg. neither in the MVVM Toolkit nor in ComputeSharp). Feels like a lot of extra work too.

> "If we just provide a code fix for this case that adds the partial modifier, it would be easy for the user to correct it."

That would be very welcome. The main issue right now for us is that as the MVVM Toolkit gets more traction (especially now that MAUI has endorsed it), we keep getting more and more people opening issues about this just because they have (understandably) no idea what `partial` even means or how source generators work at all, so even if Roslyn currently does emit an error message, they just don't really know what they're supposed to do there (or why that's even required). The result is that many of them just assume there's some bug in the library, and then go on GitHub to report it. 🥲

**Comment by CyrusNajmabadi** [comment]
> The guidance I've been giving is source generators generally do not need to provide a diagnostic for any case where the compiler is already going to report a compilation error.

I disagree.  One is an up front check of pre-reqs with clear explanations of what's going on.  THe other is a post period break without any clear link to what has happened.

I liken it to throwing an argument-exception up front, versus getting some other crash later on in program execution.

**Comment by Neme12** [comment]
I tend to agree that the source generator should also add a diagnostic if it requires something to be partial, e.g. "Methods with the [CustomGenerator] attribute must be declared in a partial type."

**Comment by sharwell** [comment]
There is no UX defined for providing a code fix for a diagnostic reported by a source generator. If you create a duplicate diagnostic, you're going to end up with one "clear but not automatically fixable" error and one "unclear but automatically fixable" error. Users will navigate to the former and not be able to quickly correct it.

**Comment by mrlacey** [comment]
> The guidance I've been giving is source generators generally do not need to provide a diagnostic for any case where the compiler is already going to report a compilation error. 

But, this issue has been opened because this guidance has turned out to be inadequate. People are not reading, or not understanding the compilation error message and so are assuming an error in the source generator.

**Comment by RikkiGibson** [comment]
In the linked issue CommunityToolkit/dotnet#286, the user said:

> I did not see anything indicating missing partial modifier when I tried to hover.

Can we verify why this happened? It feels like the "missing partial modifier" message is adequate, though it certainly could help to mention generators in the message.

**Comment by Sergio0694** [comment]
> "It feels like the "missing partial modifier" message is adequate"

To me and many others would be, but I should note that:
- Plenty (and I mean a **ton**) of devs literally do not know what "partial" means, or that it's a thing at all.
- Plenty of devs don't know that source generators literally generate new source files in their own projects.

I feel like part of the issue is that whereas this is arguably a more advanced language feature (yeah I can see that it's not that complicated, but you get what I mean), you need to consider that lots of users of source generators are really just beginners, just putting together their very first programs or applications, and this is a recurring issue for them. As I've said, I keep seeing the same exact issue being raised by early adopters of the new MVVM Toolkit source generators. If we could try to make this first experience on their part easier and better, it would both make their live easier, and also reduce churn on maintainers having to triage and explain the same thing over and over again 🙂

**Comment by CyrusNajmabadi** [comment]
Can't the SG just tell teh user "Add the 'partial' modifier to this type to enable this feature that you're trying to use."?

**Comment by Sergio0694** [comment]
Yes, and we're back at the issue about asking every single SG out there to reimplement all of this every single time.
Not to mention this would be inconsistent, because 10 different SGs will have this 10 different ways (and some not at all).
Just having the compiler emit a better message seems fine to me. It's already checking this and emitting a diagnostics.
And it clearly has all the information to determine when this is the case versus the user just defining the types themselves.

**Comment by RikkiGibson** [comment]
I would rather not put the onus on the SG authors here. I think we can solve it once and for all here. I see 2 action items for us at this point.

1. Adjust the diagnostic message to suggest adding a partial modifier in the vein of "are you missing a using directive or assembly reference?". People are probably not connecting the dots between "another partial declaration of this type exists" and "add a partial modifier here to make the error go away." We can also consider saying that the containing file appears to be from a generator, but I'm actually not sure that's *as* necessary.
2. Let's ensure we have a code fix to add the partial modifier. I produced this scenario in the IDE just now and didn't see one. 
![image](https://user-images.githubusercontent.com/5833655/171293633-622667a6-4c57-444a-ba0e-824296c00237.png)

**Comment by Neme12** [comment]
I'd be interested in adding the code fix.

**Comment by RikkiGibson** [comment]
Opened #61611 to track adding a code fix.

**Comment by CyrusNajmabadi** [comment]
> Let's ensure we have a code fix to add the partial modifier. I produced this scenario in the IDE just now and didn't see one.

In general, we've been wary of code-fixes which are effectively nothing more than adding a token (since that's what completion can do anyways).  Writing 'partial' is really not a big deal, and it's easy to do with completion anyways, so i'm not sure how having this big list to solve the issue is much better than using the other list we already offer there.

**Comment by RikkiGibson** [comment]
> Also, we should make sure that if the user hits F12 at the location of the error, they see all files defining the type.

I verified that this happens in VS and VS Code.

**Comment by Neme12** [comment]
> > Let's ensure we have a code fix to add the partial modifier. I produced this scenario in the IDE just now and didn't see one.
> 
> In general, we've been wary of code-fixes which are effectively nothing more than adding a token (since that's what completion can do anyways).  Writing 'partial' is really not a big deal, and it's easy to do with completion anyways, so i'm not sure how having this big list to solve the issue is much better than using the other list we already offer there.

The list of offered code fixes is big but that is mainly because of analyzers and code style changes that can be applied more broadly. This doesn't factor into that because this would address a compiler error that the user won't encounter that often, and it will be (or should be) the first option in the list of code fixes and should jump out at the user that there's something they can do to fix a compiler error. It won't make the list longer for the common case. There's already a few fixes that fix specific compiler errors like this, even for adding modifiers, one of which I wrote I few years back. There's also an issue somewhere about listing compiler error messages and code fixes we have for them and where we don't but could have, with the intention that we should have as many compiler errors covered to over something that fixes them.

**Comment by Neme12** [comment]
And as has been pointed out, some users might not even know about the "partial" feature and if they immediately see a light bulb with an error icon that is there to fix the issue, they can just use that right away and this can be really helpful especially to beginners.

**Comment by CyrusNajmabadi** [comment]
> And as has been pointed out, some users might not even know about the "partial" feature and if they immediately see a light bulb with an error icon that is there to fix the issue, they can just use that right away and this can be really helpful especially to beginners.

I think we coudl say that about any particular error.  But i'm not a fan of the idea of just having things "fix" the issue without users not understanding it.  That doesn't really help the user, it just keeps them in the dark about what is actually happening.  I don't think that's a good thing here.  If a user doesn't understand partial types, or what these generators are doing, that's not something to paper over on the ide side with features.  Instead, we need to solve this with education.

**Comment by CyrusNajmabadi** [comment]
> they have to go ask for help to figure it out.

I don't see this as a problem.  If users are not understanding the issue, the solution is not just to have something that they can blindly invoke and not understand.  The fix is to get them to understand.

**Comment by CyrusNajmabadi** [comment]
> It might be helpful to know which fixes already exist which essentially just add a modifier, as well as which fixes may have been suggested in the past to do just that but were rejected.

In general, if it's been a fix to just fixup a compiler error that already says waht the problem is, and it's just a token, we reject it.  For example, if you give the wrong accessibility on something, our position has been: well, just write the accessibility that fixes it.

Where we have analyzers/fixers here are for where it's often an improvement over things.  For example, if you code isn't accessing instance state, we think it's clearer for a member to be made static.  in this case though, it's not fixing some error saying "this needs to be static".

-

We've also taken this position with basic compiler errors like "hey... you missed a paren".  We don't lightbulb that, as it's more work than just typing the paren.

**Comment by Neme12** [comment]
Here's the existing issue for auditing compiler errors as candidates for code fixes by @sharwell (which implies that adding them is desirable):
https://github.com/dotnet/roslyn/issues/23326#issuecomment-425754792

Here's a few existing ones:
https://github.com/dotnet/roslyn/pull/32422
https://github.com/dotnet/roslyn/pull/30519
https://github.com/dotnet/roslyn/pull/30430

This isn't really contributing to the issue of the lightbulb menu being too cluttered at all as these kind of fixes are only offered for specific compiler errors and there will almost always be at most one fix for a compiler error and it will be at the top and really clear that this is something special, really focused and different from the other ones.

> I don't see this as a problem. If users are not understanding the issue, the solution is not just to have something that they can blindly invoke and not understand. The fix is to get them to understand.

The fix might actually help them to understand though .

**Comment by Neme12** [comment]
And users will see that there's something special there that will help them fix the error, not just regular code style changes thanks to the error lightbulb icon:
![image](https://user-images.githubusercontent.com/11444821/171297932-eea7ea17-af1f-44c1-949c-902aa43e7be2.png)

**Comment by Neme12** [comment]
And the conclusion in #32422 (https://github.com/dotnet/roslyn/pull/32422#issuecomment-454194699) was:
> @CyrusNajmabadi:
> 
> > Note: something that changes the equation here is that we are now open source. So, whereas in the past this would have been something that was potentially a large amount of cost for the team to take on, this could now be considered something that would be much less of an issue thanks to community contributions. These types of fixes would be ideal. They're a great way for people to dip their toes into roslyn, and it can allow many to 'scratch an itch' that wasn't available to them before.
> 
> To me this is entirely the game changer. If this code fix had some notable maintenance burden to it, maybe I'd be a bit concerned. But I'm not worried n this case.

**Comment by CyrusNajmabadi** [comment]
note: #23326 had the same debate here.
https://github.com/dotnet/roslyn/pull/32422 is not the same.  With that, you need to write out the actual type name, which isn't necessarily trivial.
https://github.com/dotnet/roslyn/pull/30430 is not the same as you have to determine which literal to use. etc.

These are also not cases where the completion list provides thsi information.

**Comment by CyrusNajmabadi** [comment]
I disagree with that conclusion.  The maintenance burden is there.  And we have had to maintain and bear the costs of this ourselves.  My position has been further solidified *since* that time period as the IDE team is stretched so think that we're barely keeping up with just language changes, and we have had to fix and maintain these added features, despite things being OSS.

**Comment by Neme12** [comment]
> I disagree with that conclusion. The maintenance burden is there. And we have had to maintain and bear the costs of this ourselves. My position has been further solidified _since_ that time period as the IDE team is stretched so think that we're barely keeping up with just language changes, and we have had to fix and maintain these added features, despite things being OSS.

I looked at all of those 3 features I added and all commits since I added them are automated changes and moves to a different project.

**Comment by Neme12** [comment]
I imagine that cost was due to nullable annotations to a lot of features?

**Comment by RikkiGibson** [comment]
> I don't see this as a problem. If users are not understanding the issue, the solution is not just to have something that they can blindly invoke and not understand. The fix is to get them to understand.

That is a good point. I do think that having a fixer doesn't *hurt* in this respect. If a user is so new that they don't know what *add a partial modifier* means, and the fix preview shows them a popup of what that *literally* means, that's not a bad thing. However, I can't claim that it means an *add partial* code fix meets the bar for inclusion.

Let's do what we can to make the error message easier to understand. My only remaining concern here is whether people are actually getting in a position where they are *hovering to see that error message* and the IDE isn't showing it to them. It wasn't clear to me whether that was the case. https://github.com/dotnet/roslyn/issues/61597#issuecomment-1142661339 If we can uncover an issue there, let's make sure to file and address that separately.

**Comment by sharwell** [comment]
> In general, we've been wary of code-fixes which are effectively nothing more than adding a token (since that's what completion can do anyways).

This is true for refactorings, but not true for code fixes. Analyzers (including but not limited to compiler errors) should include code fixes except in cases where the fix is unclear. This diagnostic is clear and easily fixable; there is no reason to reject a submission adding a code fix for it.

**Comment by RikkiGibson** [comment]
Let's keep @Sergio0694's points here as we consider what to do.

> * Plenty (and I mean a **ton**) of devs literally do not know what "partial" means, or that it's a thing at all.
> * Plenty of devs don't know that source generators literally generate new source files in their own projects.

I think adjusting the error message could address the second point, which will solve "wait, why is a thing called 'partial' involved in my project" :)

The first point, I really am not sure. F1 help or even Quick Info on the `partial` modifier itself when it is present?

**Comment by CyrusNajmabadi** [comment]
> This diagnostic is clear and easily fixable; there is no reason to reject a submission adding a code fix for it.

I disagree. We can see the usage numbers on these. They are a maintenance cost (along with others with equally low usage that we have had to address bugs in). 

As I said before, we are stretched extremely thin. We do not have the resources currently to address high value items and these costs do add. 

We need to focus efforts and energy (including what makes it into the code we have to own) to things with enough value.

**Comment by CyrusNajmabadi** [comment]
I genuinely don't understand the problem this is trying to solve. We already have f1 on computer diagnostics (as well as a help link) to explain what the issue is and to educate the user.  Why is that not sufficient?

I'm not seeing why we'd change course here on this issue. 

--

Note: based on telemetry numbers, there are a lot of items we're maintaining that I'll talk to mika about if we can remove. It genuinely doesn't make sense to hold onto summer of these when we don't even have dozens of uses.

**Comment by CyrusNajmabadi** [comment]
> F1 help or even Quick Info on the partial modifier itself when it is present?

F1 should work on 'partial'. If it doesn't, that's an oversight we can correct :-)

**Comment by CyrusNajmabadi** [comment]
> I think the usage might be higher if the IDE advertised code fixes for errors more in the editor, 

Many other fixes have huge usage (literally 5+ orders of magnitude more).

**Comment by Bosch-Eli-Black** [comment]
https://github.com/dotnet/roslyn/pull/65018 has been merged, so there's now a code-fix for adding the `partial` modifier 🙂
